File: python/kata02/day2.py
from kata02 import test_chop


def chop(val, seq):

    if not seq:
        return -1

    if len(seq) == 1:
        return 0 if seq[0] == val else -1

    s, e = 0, len(seq) - 1

    while e - s > 1:
        c = (e - s) / 2
        if seq[c] > val:    # left half
            s, e = s, c
        elif seq[c] < val:  # right half
            s, e = c+1, e
        else:
            return c

        if seq[s] == val:
            return s
        elif seq[e] == val:
            return e

    return -1


if __name__ == '__main__':
    test_chop(chop)

File: java/src/main/java/com/mycompany/codekata/kata02/Day2.java
package com.mycompany.codekata.kata02;

/**
 * Kata 2 from http://codekata.com/,
 *
 */
public class Day2 {

  /**
   * Binary search for value in array. Recursive version.
   *
   * @param val value to search for
   * @param arr array to search in
   * @return position of val in arr or -1 if not found
   */
  public static int chop(int val, int[] arr, int start, int length) {
    int cut;

    if (arr.length == 0)  {
      return -1;
    }

    if (length == 1) {
      return arr[start] == val ? start : -1;
    }

    cut = length / 2;

    if (arr[start + cut] <= val) {  // right
      return chop(val, arr, start + cut, length - cut);
    } else {                        // left
      return chop(val, arr, start, cut);
    }
  }

  public static int chop(int val, int[] arr) {
    return chop(val, arr, 0, arr.length);

  }
}
